### ISSUE TITLE
aws-ecs: EFSVolumeConfiguration is an extraneous key

### ISSUE THREAD

**Issue by joe-glasgow** [aws/aws-cdk#14118]
### The Question

When creating a FargateTaskDefinition and adding an EFS Volume, the CloudFormation template warns that the "EfsVolumeConfiguration" is an extraneous key - is this a known issue?

Given the following Typescript FargateTaskDefinition

      

        const myAppFargateTaskDefinition = new FargateTaskDefinition(this, 'MyAppFargateTaskDefinition', {
             memoryLimitMiB: 512,
            cpu: 256,
            executionRole: myAppFargateTaskExecutionRole,
            taskRole: myAppFargateTaskRole,
        });

        myAppFargateTaskDefinition.addVolume({
            name: 'MyAppEfsVolume',
            efsVolumeConfiguration: {
                fileSystemId: fileSystem.fileSystemId,
                transitEncryption: 'ENABLED',
                authorizationConfig: {
                    accessPointId: fileSystemAccessPoint.accessPointId
                }
            }
        });

The output Cloudformation template generates the following: 

        "MyAppFargateTaskDefinitionEB5C71CC": {
              "Type": "AWS::ECS::TaskDefinition",
              "Properties": {
                  ...,
                  Volumes": [ {
                    "Name": "MyAppEfsVolume",
                    "EfsVolumeConfiguration": {
                      "AuthorizationConfig": {
                        "AccessPointId": {
                          "Ref": "FileSystemAccessPointF8178182"
                        }
                      },
                      "TransitEncryption": "ENABLED",
                      "FileSystemId": {
                        "Ref": "FileSystem8A8E25C0"
                      }
                   }
               }]
           }
        }

However when creating the stack there is the following validation message:

`Resource template validation failed for resource MyAppFargateTaskDefinitionEB5C71CC as the template has invalid properties. Please refer to the resource documentation to fix the template. Properties validation failed for resource MyAppFargateTaskDefinitionEB5C71CC with message: #/Volumes/0: extraneous key [EfsVolumeConfiguration] is not permitted`

### Environment

  - **CDK CLI Version:** 1.97.0
  - **Module Version:** 1.97.1
  - **Node.js Version:** v13.14.0
  - **OS:** macOs Mojave
  - **Language (Version):** Typescript (4.2.4)

**Comment by joe-glasgow** [comment]
Wondering if the Cloudformation template should output 

`EFSVolumeConfiguration` instead of `EfsVolumeConfiguration` as detailed [here](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ecs-taskdefinition-volumes.html#cfn-ecs-taskdefinition-volume-efsvolumeconfiguration) ?

**Comment by peterwoodworth** [comment]
You can cdk synth and manually change and upload the output given. If you cdk synth here and manually change `EfsVolumeConfiguration` to `EFSVolumeConfiguration` and try to upload that to cloudformation will it work?

**Comment by joe-glasgow** [comment]
> You can cdk synth and manually change and upload the output given. If you cdk synth here and manually change `EfsVolumeConfiguration` to `EFSVolumeConfiguration` and try to manually upload that to cloudformation will it work?

Thanks, that did indeed fix this instance however, there is a subsequent casing issue for FileSystemId

with message: #/Volumes/0/EFSVolumeConfiguration: required key [FilesystemId] not found #/Volumes/0/EFSVolumeConfiguration: extraneous key [FileSystemId] is not permitted

Will manually change this too, will try and find the offending code later and PR a fix!

**Comment by peterwoodworth** [comment]
Great, ping me when you submit it so I can take a look!

**Comment by peterwoodworth** [comment]
No rush at all @joe-glasgow, just wondering if you're still intending on submitting a fix for this?

**Comment by jriquelme** [comment]
same problem here, waiting for that PR :)

**Comment by joe-glasgow** [comment]
Hey apologies, not had time to progress this any further really - if anyone would like to pick it up before I get around to it, I have a branch here: https://github.com/joe-glasgow/aws-cdk/compare/chore/update-efs-volume-casing?expand=1

**Comment by peterwoodworth** [comment]
I have an update after researching this more myself. Sorry for not catching this sooner @joe-glasgow 

This issue lies with CFN and our auto-generation of our L1s. It seems like this is a breaking change published in the latest version of the `cfnSpec`, we will reach out to the CFN service team to make sure it was intentional. In any case, this will get resolved when the new spec version is merged to the CDK (usually a few days)

**Comment by joe-glasgow** [comment]
Great to hear, many thanks for getting back to me on this!

**Comment by luchees** [comment]
Thanks Joe, had the same issue. For now I've manually updated the cloudformation.

**Comment by john-shaskin** [comment]
I'm still seeing this error message, after having updated to CDK `1.101.0`. Are you expecting this casing issue to be fixed in that release?

**Comment by peterwoodworth** [comment]
When this issue is closed then I'll expect the issue to be fixed. We're still waiting on Cloudformation, thank you all for your patience!

**Comment by john-shaskin** [comment]
My apologies. I fooled myself into thinking that this issue was closed--not reading the status at the top--until after I'd posted my previous comment.

**Comment by peterwoodworth** [comment]
We've released CDK version `1.102.0` with the new CfnSpec included. Let me know if the issue is fixed now @joe-glasgow @john-shaskin ! Please ping me if it still isn't fixed yet

**Comment by john-shaskin** [comment]
@peterwoodworth, I've just updated to CDK `1.102.0`, and it still appears to be rendering the ECS task definition volumes as `EfsVolumeConfiguration` rather than `EFSVolumeConfiguration`.

**Comment by peterwoodworth** [comment]
It should still successfully deploy - it deploys for me even though the synth looks the same. Are you running into any errors with your deployment @john-shaskin ?

**Comment by peterwoodworth** [comment]
```ts
    const vpc = new Vpc(this, "myvpc");

    const filsys = new FileSystem(this, "myfilesystem", {
      vpc: vpc
    });

    const taskdef = new ecs.FargateTaskDefinition(this, 'taskdef')

    taskdef.addContainer("WebContainer", {
      image: ecs.ContainerImage.fromRegistry("amazon/amazon-ecs-sample"),
    });

    taskdef.addVolume({
      name: 'myvolume',
      efsVolumeConfiguration: {
        fileSystemId: filsys.fileSystemId
      }
    })
```
minimal stack which I got to deploy

**Comment by john-shaskin** [comment]
Hey, @peterwoodworth. My stack does successfully deploy, and wasn't technically failing before. It's just a bit worrisome to see errors showing up during the deploy, and wonder if there may end up being either a CloudFormation or CDK release after which the deployment starts failing.

**Comment by peterwoodworth** [comment]
Hmmm I still can't seem to get the error to show up during deployment (which is what I was implying by getting it to deploy, sorry for not being clear).

Are you still encountering the error now? @john-shaskin

**Comment by john-shaskin** [comment]
So when you synth the template, it still renders as `EfsVolumeConfiguration`, but you don't see an error message in the Events tab of the CloudFormation stack? I should also clarify that I don't think I've seen this error in the CDK output. It's just been when looking at the AWS console.

**Comment by alexb-uk** [comment]
If it helps I'm running `1.102.0` and am still seeing the warning in CDK console out:

```
 1/3 | 12:47:34 | CREATE_IN_PROGRESS   | AWS::ECS::TaskDefinition | TaskDef (TaskDef54694570) Resource template validation failed for resource TaskDef54694570 as the template has invalid properties. Please refer to the resource documentation to fix the template.
Properties validation failed for resource TaskDef54694570 with message:
#/Volumes/0: extraneous key [EfsVolumeConfiguration] is not permitted
```

Synth template:
```json
"Volumes": [
    {
      "Name": "efs-test",
      "EfsVolumeConfiguration": {
        "FileSystemId": {
          "Ref": "EfsFileSystemXXX"
        }
      }
    },
```

Thanks

**Comment by alexb-uk** [comment]
To answer your other question yes I do see the warning message in CF Events:
![image](https://user-images.githubusercontent.com/670824/117687742-fc95c900-b1af-11eb-9603-824501996a68.png)

**Comment by iliapolo** [comment]
> We have a [patch in place](https://github.com/aws/aws-cdk/blob/7966f8d48c4bff26beb22856d289f9d0c7e7081d/packages/%40aws-cdk/cfnspec/spec-source/620_Ecs_TaskDefinition_EfsVolumeConfiguration_patch.json) to keep the casing this way.
> 
> @iliapolo it looks like you made this patch to prevent any breaking changes from occurring. Is this patch still necessary to prevent a breaking change?

@peterwoodworth I'm not sure. We applied to patch to avoid resource replacement. So if it doesn't cause a replacement, It can be removed.

**Comment by MrArnoldPalmer** [comment]
Hey everyone, check out [this comment here](https://github.com/aws/aws-cdk/issues/15025#issuecomment-858233367) for an explanation. I'm closing this in favor of #15205 since they are the same issue. Add comments/responses there please 🙏🏻
